### DISCUSSION TITLE
urql/svelte v2 mutations are missing explicit execute function

### DISCUSSION THREAD

**Discussion by freb** [urql-graphql/urql#2499] [Q&A] [answered]
I've been using the `@urql/svelte` v2 code since it was released and quickly ran into an issue with mutations, which I asked about here: https://github.com/FormidableLabs/urql/discussions/2479

I initially assumed I just didn't understand the code and needed some guidance. But after spending more time with it, I think that the `mutationStore` implementation is lacking functionality to make it useful.

Currently, mutationStore runs as soon as it is called and returns an `OperationResultStore`.  The issue is that it seems impossible to create a mustationStore without either (1) running an initial mutation with variables that are not yet known, or (2) creating a mutation that cannot be subscribed to.

For example, let's suppose you want to update a user profile. Before you submit a mutation, you first need to get the values from input fields, such a `firstName`. However, if you wait to create the `mutationStore` until an input field has changed, you will be creating the `mutationStore` inside of a callback function which cannot be subscribed to the from the Svelte markup. Similarly, if you create the `mutationStore` with placeholder variables so that the store can be accessed from the Svelte markup, an initial mutation request will be made to update their values before they have been properly set.

In either case, `mutationStore` will need to be wrapped and have its functionality duplicated in order to subscribe to it in the markup but defer its execution (which I haven't figured out yet).

**urql version & exchanges**

~~~
├── @urql/devtools@2.0.3
├── @urql/exchange-auth@0.1.7
├── @urql/exchange-graphcache@4.4.3
├── @urql/introspection@0.3.2
├── @urql/svelte@2.0.1
~~~

**Steps to reproduce**

Import `mutationStore` from `@urql/svelte` (v2) and attempt to use it to submit form data where you can subscribe to it and where it does not make an undesired initial request.

**Expected behavior**

I don't know how the functional API design should differ from the v1 API for mutations, but if the implementation is to look as close as possible to what we have now, it should at least have the option to initialize in a paused state to be explicitly unpaused later, after the variables have are known and have been set.

**Actual behavior**

A `mutationStore` is created and either (1) cannot be subscribed to or (2) submits an initial request using variable values before they are known.

**Comment by frederikhors** [comment]
I'm on the same boat. V2 is not production ready unfortunately.

**Comment by JoviDeCroock** [comment, marked as answer]
the v2 API moves more towards the functional paradigm enforced in Svelte hence we are now not providing an execute function rather than the store being [the executor](https://formidable-com-urql-staging-2501.surge.sh/open-source/urql/docs/basics/svelte/#sending-a-mutation). I am personally no Svelte expert and I don't actively develop with it but I do think this is more in line with how they expect things to run, please correct me if I am off base here 😅

**Reply by freb** [reply]
@JoviDeCroock, thanks for a link to the updated document. Looks like I just needed a little more patience since the example there pretty much answered my mutation questions.

Here is the current version of my code based on the example, which is working great but probably needs some more cleanup:

~~~svelte
<script lang="ts">
let firstName: string
let id = "abc123"
let result: OperationResultStore<UsersWhereQuery, UpdateUserMutationVariables> | null = null

let updateUserMut = (variables: UpdateUserMutationVariables) => {
  result = mutationStore<UsersWhereQuery, UsersWhereQuery, UpdateUserMutationVariables>({
    client: $client,
    query: updateUserMutation,
    variables,
  })
}

function handleInput(e: InputEvent) {
  const { target } = e
  if (target) {
    firstName = (target as HTMLInputElement).value
  }
}

function handleSave() {
  updateUserMut({id, input: {firstName}})
}
</script>

firstName (variable): {firstName}<br/>
firstName (mut result): {#if result && $result.data}{$result.data.updateUser.firstName}{:else}NOT READY{/if}<br/>

<input value="get initial value from separate queryStore" on:input={handleInput}/>
<button on:click={handleSave}>SAVE</button><br>
~~~

**Reply by mpiorowski** [reply]
@freb how did you manage to create client store? ($client)

**Reply by freb** [reply]
You can see what I am doing in this example I created to demonstrate an issue: https://stackblitz.com/edit/vitejs-vite-2ssvxj?file=src/routes/index.svelte

I'm basically just creating a readable store with the client instead of pulling it from context. I think either method is fine. At least this method does not seem to be the cause of my other problems.

**Reply by mpiorowski** [reply]
@freb Have you manage to make mutation automaticlly refetch query? Cos for me it's not working :(

**Reply by freb** [reply]
@mpiorowski I've gotten much more familiar with urql/svelte over the last few weeks. I missed your comment before but I might be able to help if you're still having an issue.

Also, I ran into issues clearing the cache (creating and using a new client) with the code linked to above. I needed to switch to a writable store so that I could swap the client out for a new one on logout. Here is the current version:

~~~ts
import { createClient } from '@urql/svelte';
import { writable } from 'svelte/store';


const newClient = () => {
	return createClient({
		// Your parameters here.
	})
}

function createStore() {
	const { subscribe, set } = writable(newClient());

	return {
		subscribe,
		clearCache: () => set(newClient()),
	}
}

export default createStore()
~~~

**Comment by SiNONiMiTY** [comment]
@freb how did you import the OperationResultStore type? I am hitting the same issue that you originally have, and even the sample mutation on the documentation does not work for me.

The problem I am encountering is exactly as the original post, as soon as the page loads, the useMutation() function is immediately called, therefore a, blank request is sent. 

If I can't find any resolution with this, I think the fetch() API will suffice for querying the GraphQL endpoints.

**Reply by freb** [reply]
I don't think you can actually import the `OperationalResultStore`. If you're looking to do something like what I have in my example code, the best you can do is use the `ReturnType` of `queryStore`. Something like:

~~~ts
import {
	queryStore,
	type AnyVariables,
} from "@urql/svelte";

type OperationalResultStore<Data, Variables extends AnyVariables> = ReturnType<typeof queryStore<Data, Variables>>
~~~

But the actual solution I went with is simply to not use stores for mutations since I don't think they make sense in most situations. Just separate the display portion of the code from the update portion of the code. Use a `queryStore` to fetch the information you want to display and then use `client.mutation` to make updates by calling it inside of the form submit handler, for example. With proper caching the store's variable being displayed will be immediately updated.

I use the `.toPromise()` method on `client.mutation` inside an async call so that I can handle any errors that arise before returning from the form handler.

**Reply by SiNONiMiTY** [reply]
Thanks @freb for the insight, will try the client.mutation() API and see if it suits my use case better
